Mathlib/Algebra/Lie/Weights/Chain.lean

```diff
@@ -0,0 +1,130 @@
+
+/-- Given two (potential) weights `χ₁` and `χ₂` together with integers `p` and `q`, it is often
+useful to study the sum of weight spaces associated to the family of weights `k • χ₁ + χ₂` for
+`p < k < q`. -/
+def weightSpaceChain : LieSubmodule R L M :=
+  ⨆ k ∈ Ioo p q, weightSpace M (k • χ₁ + χ₂)
+
+lemma weightSpaceChain_def :
+    weightSpaceChain M χ₁ χ₂ p q = ⨆ k ∈ Ioo p q, weightSpace M (k • χ₁ + χ₂) :=
+  rfl
+
+lemma weightSpaceChain_def' :
+    weightSpaceChain M χ₁ χ₂ p q = ⨆ k ∈ Finset.Ioo p q, weightSpace M (k • χ₁ + χ₂) := by
+  have : ∀ (k : ℤ), k ∈ Ioo p q ↔ k ∈ Finset.Ioo p q := by simp
+  simp_rw [weightSpaceChain_def, this]
+
+@[simp]
+lemma weightSpaceChain_neg :
+    weightSpaceChain M (-χ₁) χ₂ (-q) (-p) = weightSpaceChain M χ₁ χ₂ p q := by
+  let e : ℤ ≃ ℤ := neg_involutive.toPerm
+  simp_rw [weightSpaceChain, ← e.biSup_comp (Ioo p q)]
+  simp [-mem_Ioo, neg_mem_Ioo_iff]
+
+end IsNilpotent
+
+section IsNilpotentSubalgebra
+
+open LieAlgebra
+
+variable {H : LieSubalgebra R L} (α χ : H → R) (p q : ℤ)
+
+lemma lie_mem_weightSpaceChain_of_weightSpace_eq_bot_right [LieAlgebra.IsNilpotent R H]
+    (hq : weightSpace M (q • α + χ) = ⊥)
+    {x : L} (hx : x ∈ rootSpace H α)
+    {y : M} (hy : y ∈ weightSpaceChain M α χ p q) :
+    ⁅x, y⁆ ∈ weightSpaceChain M α χ p q := by
+  rw [weightSpaceChain, iSup_subtype'] at hy
+  induction' hy using LieSubmodule.iSup_induction' with k z hz z₁ z₂ _ _ hz₁ hz₂
+  · obtain ⟨k, hk⟩ := k
+    suffices weightSpace M ((k + 1) • α + χ) ≤ weightSpaceChain M α χ p q by
+      apply this
+      simpa using (rootSpaceWeightSpaceProduct R L H M α (k • α + χ) ((k + 1) • α + χ)
+        (by rw [add_smul]; abel) (⟨x, hx⟩ ⊗ₜ ⟨z, hz⟩)).property
+    rw [weightSpaceChain]
+    rcases eq_or_ne (k + 1) q with rfl | hk'; · simp only [hq, bot_le]
+    replace hk' : k + 1 ∈ Ioo p q := ⟨by linarith [hk.1], lt_of_le_of_ne hk.2 hk'⟩
+    exact le_biSup (fun k ↦ weightSpace M (k • α + χ)) hk'
+  · simp
+  · rw [lie_add]
+    exact add_mem hz₁ hz₂
+
+lemma lie_mem_weightSpaceChain_of_weightSpace_eq_bot_left [LieAlgebra.IsNilpotent R H]
+    (hp : weightSpace M (p • α + χ) = ⊥)
+    {x : L} (hx : x ∈ rootSpace H (-α))
+    {y : M} (hy : y ∈ weightSpaceChain M α χ p q) :
+    ⁅x, y⁆ ∈ weightSpaceChain M α χ p q := by
+  replace hp : weightSpace M ((-p) • (-α) + χ) = ⊥ := by rwa [smul_neg, neg_smul, neg_neg]
+  rw [← weightSpaceChain_neg] at hy ⊢
+  exact lie_mem_weightSpaceChain_of_weightSpace_eq_bot_right M (-α) χ (-q) (-p) hp hx hy
+
+lemma trace_toEndomorphism_weightSpaceChain_eq_zero [IsNoetherian R L] [H.IsCartanSubalgebra]
+    (hp : weightSpace M (p • α + χ) = ⊥)
+    (hq : weightSpace M (q • α + χ) = ⊥)
+    {x : H} (hx : x ∈ (rootSpaceProductNegSelf α).range) :
+    LinearMap.trace R _ (toEndomorphism R H (weightSpaceChain M α χ p q) x) = 0 := by
+  obtain ⟨t, rfl⟩ := hx
+  induction' t using TensorProduct.induction_on with y z _ _ h₁ h₂
+  · simp
+  · let f : Module.End R (weightSpaceChain M α χ p q) :=
```

Comment:
is this explciit characterisation likely to be useful outside the proof? I guess probably not as we're only dealing with pure elements, right?

Reply:
Yes, I think it unlikely these will be useful outside.